### PR TITLE
slot2ssa: consider liveness when inserting `PiNode`s

### LINKED ISSUE #51144: Internal compiler error: UndefRefError during compaction
MWE, reduced from package using Automa.jl (regex code generator):

```julia
function f(x, y)
    if if x
        if if if if if if if if if if if if if if if if if if if @goto bar
        end end end end end end end end end end end end end end end end end end end
    end end
    @label foo
    if y end
    @label bar
    if (y || 0) && @goto foo
    end
end

Base.code_ircode(f, (Bool, Bool))
```

```
$ manyjulias --asserts master wip.jl                  
[ Info: Translated requested revision master to commit 631b5c39b72d91bec33262e14ce96f74de9aa7b6
ERROR: LoadError: UndefRefError: access to undefined reference
Stacktrace:
  [1] getindex
    @ Core.Compiler ./essentials.jl:13 [inlined]
  [2] getindex
    @ Core.Compiler ./compiler/ssair/ir.jl:244 [inlined]
  [3] process_node!(compact::Core.Compiler.IncrementalCompact, result_idx::Int64, inst::Core.Compiler.Instruction, idx::Int64, processed_idx::Int64, active_bb::Int64, do_rename_ssa::Bool)
    @ Core.Compiler ./compiler/ssair/ir.jl:1361
  [4] process_newnode!(compact::Core.Compiler.IncrementalCompact, new_idx::Int64, new_node_entry::Core.Compiler.Instruction, new_node_info::Core.Compiler.NewNodeInfo, idx::Int64, active_bb::Int64, do_rename_ssa::Bool)
    @ Core.Compiler ./compiler/ssair/ir.jl:1543
  [5] iterate_compact(compact::Core.Compiler.IncrementalCompact)
    @ Core.Compiler ./compiler/ssair/ir.jl:1654
  [6] iterate
    @ Core.Compiler ./compiler/ssair/ir.jl:1597 [inlined]
  [7] compact!(code::Core.Compiler.IRCode, allow_cfg_transforms::Bool)
    @ Core.Compiler ./compiler/ssair/ir.jl:1862
  [8] compact!
    @ Core.Compiler ./compiler/ssair/ir.jl:1860 [inlined]
  [9] run_passes(ci::Core.CodeInfo, sv::Core.Compiler.OptimizationState{Core.Compiler.NativeInterpreter}, caller::Core.Compiler.InferenceResult, optimize_until::Nothing)
    @ Core.Compiler ./compiler/optimize.jl:494
 [10] typeinf_ircode(interp::Core.Compiler.NativeInterpreter, method::Method, atype::Any, sparams::Core.SimpleVector, optimize_until::Nothing)
    @ Core.Compiler ./compiler/typeinfer.jl:922
 [11] code_ircode_by_type(tt::Type; world::UInt64, interp::Core.Compiler.NativeInterpreter, optimize_until::Nothing)
    @ Base ./reflection.jl:1511
 [12] code_ircode_by_type
    @ Base ./reflection.jl:1497 [inlined]
 [13] code_ircode(f::Any, types::Any; kwargs::@Kwargs{})
    @ Base ./reflection.jl:1488
 [14] code_ircode(f::Any, types::Any)
    @ Base ./reflection.jl:1483
 [15] top-level scope
    @ ~/Julia/tools/creduce/main.jl:13
```

Bisected to #50943

Occurs on PkgEval, https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2023-08/31/BioStockholm.primary.log

### PR DESCRIPTION
**Pull request by topolarity**
If we blindly insert a `PiNode` without considering liveness, that can create a usage "out of thin air" that our `PhiNode` insertion did not account for. The result is a reference to a non-dominating value, producing invalid IR.

Resolves #51144.

p.s. I'd like to re-factor this code in the future to walk the dominator tree so that we can re-use `PiNode`'s across basic blocks and avoid some of these temporaries, but this is good as a quick fix.

### CODE REVIEW

**Review by maleadt — APPROVED** [review]
Not familiar with the code, but it does seem to fix both the MWE and BioStockholm.jl, so SGTM.

File: base/compiler/ssair/slot2ssa.jl [hunk 1]
```diff
@@ -617,7 +618,11 @@ function construct_ssa!(ci::CodeInfo, ir::IRCode, sv::OptimizationState,
             end
             continue
         end
+
         @timeit "liveness" (live = compute_live_ins(cfg, slot))
+        for block in live.live_in_bbs
+            push!(live_slots[block], idx)
+        end
         for li in live.live_in_bbs
             cidx = findfirst(x::TryCatchRegion->x.leave_block==li, catch_entry_blocks)
```
Comment (aviatesk) [inline base/compiler/ssair/slot2ssa.jl:627]:
```suggestion
        for li in live.live_in_bbs
            push!(live_slots[li], idx)
            cidx = findfirst(x::TryCatchRegion->x.leave_block==li, catch_entry_blocks)
```
maybe merge these two loops?

**Review by aviatesk — APPROVED** [review]
LGTM.

@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by topolarity** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51406#issuecomment-1730475227) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/30a6b6a_vs_57e70c1/report.md).

**Pull request merged**